homeassistant/components/sonos/number.py

```diff
@@ -142,3 +164,240 @@ def native_value(self) -> float:
+        level = int(round(max(0.0, min(100.0, float(value)))))
+
+        if self._is_grouped():
+            # Set at the coordinator
+            try:
+                coord = self._coordinator_soco()
+                coord.group.volume = level  # native Sonos GroupRenderingControl
+            finally:
+                # After writing, trigger one fresh read to fan-out the true value
+                self.hass.loop.call_soon_threadsafe(
+                    self.hass.async_create_task, self._async_refresh_from_device()
+                )
+        else:
+            # Not grouped → act as player volume mirror
+            self.soco.volume = level
+            # Write-through to UI immediately
+            self._value = level
+            self.hass.loop.call_soon_threadsafe(self.async_write_ha_state)
+
+    async def _async_fallback_poll(self) -> None:
+        await self._async_refresh_from_device()
+
+    async def _async_refresh_from_device(self) -> None:
+        """Read the *native* volume (group or player) and propagate to peers."""
+        gid_actual = self.soco.group.uid
+
+        def _get() -> int | None:
+            try:
+                if self._is_grouped():
+                    # Any member can query; SoCo will address the coordinator
+                    return int(self.soco.group.volume)
+                # Ungrouped: mirror player volume
+                return int(self.soco.volume)
+            except (SoCoException, OSError) as err:
+                _LOGGER.debug(
+                    "Failed to read volume for %s: %s", self.speaker.zone_name, err
+                )
+                return None
+
+        vol = await self.hass.async_add_executor_job(_get)
+        if vol is None:
+            return
+
+        changed = self._value != vol
+        self._value = vol
+        self.async_write_ha_state()
+
+        # Fan-out the fresh *group* value so peers update immediately
+        if self._is_grouped() and gid_actual:
+            async_dispatcher_send(self.hass, _gv_signal(gid_actual), vol)
+            if changed:
+                _LOGGER.debug(
+                    "GV refresh: zone=%s grouped=True gid=%s vol=%s",
+                    self.speaker.zone_name,
+                    gid_actual,
+                    vol,
+                )
+        elif changed:
+            _LOGGER.debug(
+                "GV refresh: zone=%s grouped=False vol=%s",
+                self.speaker.zone_name,
+                vol,
+            )
+
+    # ------------------ wiring & lifecycle ----------------
+
+    async def async_added_to_hass(self) -> None:
+        """Finish setup: bind signals and perform an initial refresh."""
+        await super().async_added_to_hass()
+
+        self._coord_uid = (self.speaker.coordinator or self.speaker).uid
+        self._group_uid = self._current_group_uid()
+
+        # Any activity from any speaker → check if topology changed, refresh once
```

Comment:
A consideration. 

Whenever group volume is changed the speaker sends a RenderingControl event which is received by speaker.py:async_update_volume.

[speaker.py](https://github.com/themegamind/core/blob/70008c191ac8e50bd53db4feb1b48407d196b564/homeassistant/components/sonos/speaker.py#L566-L568)

When the volume changes for a speaker; send an event with the coordinator uid (e.g.  SONOS_VOLUME_UPDATE-{coord_uid}) and have each number entity subscribed only to this event and then retrieve the group volume from the coordinator soco. This may fix the tracking issue reported and reduce activity fetching group volumes whenever there is unrelated activity.